File: desktop/packages/mullvad-vpn/src/renderer/lib/notifications/warren-notice.ts
import { WarrenNotice } from '../../../shared/daemon-rpc-types';
import { messages } from '../../../shared/gettext';
import { InAppNotification, InAppNotificationProvider } from '../../../shared/notifications';

interface WarrenNoticeNotificationContext {
  // Notices the daemon fetched from the API, verified against the pinned
  // server key, and filtered for expiry and this app's version. Already
  // display-ready: the renderer shows them verbatim.
  notices: WarrenNotice[];
  // Keys of the notices this user has put away, from the GUI settings.
  dismissedKeys: string[];
  dismiss: (key: string) => void;
}

/**
 * Key a dismissal is recorded under. The wording is part of it, not just the
 * id: an operator who rewrites a notice in place keeps its id, and a key on
 * the id alone would bury the new words for everyone who had put the old ones
 * away. Any stable digest does, this one is the classic djb2.
 */
export function noticeDismissalKey(notice: WarrenNotice): string {
  let hash = 5381;
  for (let i = 0; i < notice.message.length; i++) {
    hash = ((hash << 5) + hash + notice.message.charCodeAt(i)) | 0;
  }
  return `${notice.id}:${hash}`;
}

// Operator broadcast banner. Ranked FIRST among the in-app notification
// providers, so an active notice outranks every connection-state banner:
// when the operator has something to say to everyone, that is the one
// thing the user must see, and the states it hides (connecting, offline,
// error) are already visible in the connect view's own status.
//
// It clears from the signal that raised it: the daemon pushes an empty list
// when the notice is erased or lapses, so there is no renderer-side timer.
//
// An informational notice can also be put away by the reader, and only that
// one: ranked on top of a slot that holds one card, a message the operator
// leaves up for a week would otherwise hide the update prompt and the expiry
// warning for that whole week. A warning or an error keeps the slot, because
// it describes something live that the user cannot act on by hiding it.
export class WarrenNoticeNotificationProvider implements InAppNotificationProvider {
  public constructor(private context: WarrenNoticeNotificationContext) {}

  public mayDisplay = () => this.displayable() !== undefined;

  public getInAppNotification(): InAppNotification {
    const notice = this.displayable()!;
    const title = titleFor(notice.level);
    return {
      indicator: indicatorFor(notice.level),
      // The label is translated; the message body is the operator's own
      // words, shown as authored.
      title,
      subtitle: notice.message,
      // A notice can be as long as the publication cap allows, so the banner
      // clamps it and offers the full text in a scrollable modal. The banner
      // renders the affordance only when the text actually overflows.
      action: {
        type: 'expand-text',
        expand: { title, content: notice.message },
        dismiss:
          notice.level === 'info'
            ? () => this.context.dismiss(noticeDismissalKey(notice))
            : undefined,
      },
    };
  }

  private displayable(): WarrenNotice | undefined {
    return this.context.notices.find(
      (notice) =>
        notice.level !== 'info' || !this.context.dismissedKeys.includes(noticeDismissalKey(notice)),
    );
  }
}

function indicatorFor(level: WarrenNotice['level']): InAppNotification['indicator'] {
  switch (level) {
    case 'error':
      return 'error';
    case 'warning':
      return 'warning';
    default:
      return 'success';
  }
}

function titleFor(level: WarrenNotice['level']): string {
  switch (level) {
    case 'error':
      return messages.pgettext('in-app-notifications', 'IMPORTANT');
    case 'warning':
      return messages.pgettext('in-app-notifications', 'NOTICE');
    default:
      return messages.pgettext('in-app-notifications', 'WARREN');
  }
}

File: android/lib/model/src/main/kotlin/com/warrenbrowse/vpn/lib/model/WarrenNotice.kt
package com.warrenbrowse.vpn.lib.model

/**
 * One operator broadcast notice, as Rust hands it over: already verified
 * against the pinned server key and already filtered for the envelope expiry,
 * the notice's own TTL and this app's version.
 *
 * [message] is the operator's own words and is rendered verbatim as plain
 * text. It is never translated and never parsed as markup: the whole point of
 * the signed channel is that what the operator wrote is what the user reads.
 */
data class WarrenNotice(val id: String, val message: String, val level: WarrenNoticeLevel) {
    /**
     * Key a dismissal is recorded under. The wording is part of it, not just
     * the id: an operator who rewrites a notice in place keeps its id, and a
     * key on the id alone would bury the new words for everyone who had put
     * the old ones away.
     */
    val dismissalKey: String
        get() = "$id:${message.hashCode()}"
}

/** Severity of a [WarrenNotice], which picks the banner's title and colour. */
enum class WarrenNoticeLevel {
    INFO,
    WARNING,
    ERROR;

    companion object {
        /**
         * The level for a wire token, [INFO] for anything this build does not
         * know. A newer severity must degrade to the calmest banner rather
         * than drop the message: the operator still has something to say.
         */
        fun of(token: String): WarrenNoticeLevel =
            when (token) {
                "error" -> ERROR
                "warning" -> WARNING
                else -> INFO
            }
    }
}
